File: longest-palindromic-substring/longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        if not s:
            return s
        
        answer = (0, 0) # length and start tuple will compare first element as default
        for mid in range(len(s)):
            answer = max(answer, self.getString(s, mid, mid))
            answer = max(answer, self.getString(s, mid, mid + 1))
        
        return s[answer[1]: answer[0] + answer[1]]  

    def getString(self, s, left, right):
        while left >= 0 and right < len(s) and s[left] == s[right]:
            left -= 1
            right += 1
        return right - left - 1, left + 1

File: longest-palindromic-substring/longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        if (s.length() == 0) return "";
        int length = s.size();
        int maxLength = INT_MIN;
        int start = 0;
        for (int i = 0; i < s.size(); i++) {  
            checkString(s, i, i, maxLength, start);
            checkString(s, i, i + 1, maxLength, start);
        }
        return maxLength == INT_MIN? "" : s.substr(start, maxLength);
    }
private: 
        void checkString(string s, int left, int right, int& maxLength, int& start) {
            while (left >= 0 && right < s.size() && s[left] == s[right]) {
                left--;
                right++;
            }
            if (right - left - 1 > maxLength) {
                maxLength = right - left - 1;
                start = left + 1; // left左移过一次
            }
        }

};